File: mobile/lib/components/fiche_terrain.dart
import 'package:flutter/material.dart';

class FicheTerrain extends StatefulWidget {
  final Map<String, dynamic> object;
  final String type;
  final VoidCallback onClose;
  final void Function(Map<String, dynamic>) onSave;

  const FicheTerrain({
    super.key,
    required this.object,
    required this.type,
    required this.onClose,
    required this.onSave,
  });

  @override
  State<FicheTerrain> createState() => _FicheTerrainState();
}

class _FicheTerrainState extends State<FicheTerrain> {
  String etat = 'Bon';
  String niveauEau = 'Normal';
  String obstruction = 'Aucune';
  String odeur = 'Aucune';
  bool debordement = false;
  String commentaires = '';
  List<String> photos = [];
  String statut = 'À vérifier';

  void handleAddPhoto() {
    final mockPhotos = [
      'https://images.unsplash.com/photo-1542060748-10c28b629f6f?w=400',
      'https://images.unsplash.com/photo-1584622650111-993a426fbf0a?w=400',
      'https://images.unsplash.com/photo-1504307651254-35680f356dfd?w=400',
    ];
    final randomPhoto = mockPhotos[DateTime.now().millisecond % mockPhotos.length];
    final newPhoto = '$randomPhoto&sig=${DateTime.now().millisecondsSinceEpoch}';
    setState(() => photos.add(newPhoto));
  }

  void handleSave() {
    final inspectionData = {
      'object_type': widget.type,
      'object_id': widget.object['code'] ?? widget.object['fid'] ?? widget.object['id']?.toString(),
      'etat': etat,
      'niveau_eau': niveauEau,
      'obstruction': obstruction,
      'odeur': odeur,
      'debordement': debordement,
      'commentaires': commentaires,
      'photos': photos,
      'statut': statut,
      'date_inspection': DateTime.now().toIso8601String(),
    };
    widget.onSave(inspectionData);
    ScaffoldMessenger.of(context).showSnackBar(
      const SnackBar(content: Text('Inspection enregistrée avec succès !')),
    );
  }

  Widget renderDropdown(String label, String current, List<String> options, ValueChanged<String> setter) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Text(
          label,
          style: const TextStyle(
            fontSize: 13,
            fontWeight: FontWeight.bold,
          ),
        ),
        const SizedBox(height: 4),
        Wrap(
          spacing: 6,
          children: options
              .map((opt) => ChoiceChip(
                    label: Text(opt),
                    selected: current == opt,
                    onSelected: (selected) {
                      if (selected) setter(opt);
                    },
                    selectedColor: const Color(0xFF16213e),
                    labelStyle: TextStyle(
                      color: current == opt ? Colors.white : Colors.black,
                    ),
                  ))
              .toList(),
        ),
        const SizedBox(height: 12),
      ],
    );
  }

  @override
  Widget build(BuildContext context) {
    return SingleChildScrollView(
      padding: const EdgeInsets.all(12),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: [
          Container(
            color: const Color(0xFF16213e),
            padding: const EdgeInsets.all(16),
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Text(
                  'Fiche Terrain - ${widget.type.toUpperCase()}',
                  style: const TextStyle(
                    color: Colors.white,
                    fontSize: 18,
                    fontWeight: FontWeight.bold,
                  ),
                ),
                const SizedBox(height: 4),
                Text(
                  'ID : ${widget.object['code'] ?? widget.object['fid'] ?? widget.object['id']}',
                  style: const TextStyle(color: Colors.white70),
                ),
              ],
            ),
          ),
          const SizedBox(height: 12),
          Card(
            child: Padding(
              padding: const EdgeInsets.all(16),
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  const Text(
                    'Fiche Technique',
                    style: TextStyle(
                      fontSize: 15,
                      fontWeight: FontWeight.bold,
                    ),
                  ),
                  const SizedBox(height: 10),
                  if (widget.object['diametre'] != null)
                    Text('• Diamètre : ${widget.object['diametre']} mm'),
                  if (widget.object['longueur'] != null)
                    Text('• Longueur : ${widget.object['longueur']} m'),
                  if (widget.object['materiau'] != null)
                    Text('• Matériau : ${widget.object['materiau']}'),
                  if (widget.object['profondeur'] != null)
                    Text('• Profondeur : ${widget.object['profondeur']} m'),
                  if (widget.object['commune'] != null)
                    Text('• Commune : ${widget.object['commune']}'),
                  if (widget.object['nom_voie'] != null)
                    Text('• Voie : ${widget.object['nom_voie']}'),
                ],
              ),
            ),
          ),
          const SizedBox(height: 12),
          Card(
            child: Padding(
              padding: const EdgeInsets.all(16),
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  const Text(
                    'Observations terrain',
                    style: TextStyle(
                      fontSize: 15,
                      fontWeight: FontWeight.bold,
                    ),
                  ),
                  const SizedBox(height: 12),
                  renderDropdown('État général', etat, ['Bon', 'Moyen', 'Mauvais'], (v) => setState(() => etat = v)),
                  renderDropdown('Niveau d\'eau', niveauEau, ['Vide', 'Normal', 'Sous charge', 'Débordement'], (v) => setState(() => niveauEau = v)),
                  renderDropdown('Obstruction', obstruction, ['Aucune', 'Partielle', 'Totale'], (v) => setState(() => obstruction = v)),
                  renderDropdown('Odeur', odeur, ['Aucune', 'Légère', 'Forte'], (v) => setState(() => odeur = v)),
                  Row(
                    mainAxisAlignment: MainAxisAlignment.spaceBetween,
                    children: [
                      const Text('Débordement actif :', style: TextStyle(fontSize: 13)),
                      Switch(
                        value: debordement,
                        onChanged: (v) => setState(() => debordement = v),
                        activeColor: const Color(0xFFe74c3c),
                      ),
                    ],
                  ),
                  renderDropdown('Statut de validation', statut, ['À vérifier', 'Corrigé terrain', 'Validé bureau'], (v) => setState(() => statut = v)),
                  const SizedBox(height: 8),
                  const Text('Commentaires / Diagnostic :', style: TextStyle(fontSize: 13)),
                  TextFormField(
                    maxLines: 3,
                    initialValue: commentaires,
                    onChanged: (v) => commentaires = v,
                    decoration: const InputDecoration(
                      border: OutlineInputBorder(),
                      hintText: 'Renseigner les constatations terrain...',
                    ),
                  ),
                  const SizedBox(height: 12),
                  Text('Photos d\'inspection (${photos.length})'),
                  const SizedBox(height: 8),
                  Wrap(
                    spacing: 8,
                    children: [
                      ...photos.map((uri) => Image.network(
                            uri,
                            width: 50,
                            height: 50,
                            fit: BoxFit.cover,
                          )),
                      GestureDetector(
                        onTap: handleAddPhoto,
                        child: Container(
                          width: 50,
                          height: 50,
                          decoration: BoxDecoration(
                            border: Border.all(style: BorderStyle.solid),
                            borderRadius: BorderRadius.circular(4),
                          ),
                          child: const Center(child: Text('📸 +')),
                        ),
                      ),
                    ],
                  ),
                ],
              ),
            ),
          ),
          const SizedBox(height: 16),
          Row(
            children: [
              Expanded(
                child: OutlinedButton(
                  onPressed: widget.onClose,
                  child: const Text('Annuler'),
                ),
              ),
              const SizedBox(width: 8),
              Expanded(
                child: ElevatedButton(
                  onPressed: handleSave,
                  style: ElevatedButton.styleFrom(
                    backgroundColor: const Color(0xFF27ae60),
                  ),
                  child: const Text('Enregistrer'),
                ),
              ),
            ],
          ),
          const SizedBox(height: 40),
        ],
      ),
    );
  }

File: mobile/components/FicheTerrain.js
/**
 * Nom Auteur : Dr Abdelhakim Kellouche
 * Nom de l'application : Plateforme de Pilotage du Réseau d'Assainissement
 * Numéro version : 1.0.0
 * Date de création : 30-05-2026
 * Date de modification : 30-05-2026
 * 
 * Objectif du module :
 * Composant de formulaire d'inspection terrain et fiche technique unifiée pour 
 * les regards, conduites, stations de relevage et ouvrages spéciaux.
 */

import React, { useState } from 'react';
import {
  View,
  Text,
  StyleSheet,
  TouchableOpacity,
  TextInput,
  ScrollView,
  Switch,
  Alert,
  Image,
} from 'react-native';

export default function FicheTerrain({ object, type, onSave, onClose }) {
  const [etat, setEtat] = useState('Bon');
  const [niveauEau, setNiveauEau] = useState('Normal');
  const [obstruction, setObstruction] = useState('Aucune');
  const [odeur, setOdeur] = useState('Aucune');
  const [debordement, setDebordement] = useState(false);
  const [commentaires, setCommentaires] = useState('');
  const [photos, setPhotos] = useState([]);
  const [statut, setStatut] = useState('À vérifier');

  // Simulation d'ajout de photo réaliste d'inspection
  const handleAddPhoto = () => {
    const mockPhotos = [
      'https://images.unsplash.com/photo-1542060748-10c28b629f6f?w=400', // canalisation
      'https://images.unsplash.com/photo-1584622650111-993a426fbf0a?w=400', // regard/plaque
      'https://images.unsplash.com/photo-1504307651254-35680f356dfd?w=400', // travaux/terrain
    ];
    const randomPhoto = mockPhotos[Math.floor(Math.random() * mockPhotos.length)];
    const newPhoto = `${randomPhoto}&sig=${Date.now()}`;
    setPhotos([...photos, newPhoto]);
  };

  const handleSave = () => {
    const inspectionData = {
      object_type: type,
      object_id: object.code || object.fid || object.id?.toString(),
      etat,
      niveau_eau: niveauEau,
      obstruction,
      odeur,
      debordement,
      commentaires,
      photos,
      statut,
      date_inspection: new Date().toISOString(),
    };
    onSave(inspectionData);
    Alert.alert('Succès', 'Inspection enregistrée avec succès !');
  };

  const renderDropdown = (label, current, options, setter) => (
    <View style={styles.formGroup}>
      <Text style={styles.label}>{label}</Text>
      <View style={styles.optionContainer}>
        {options.map((opt) => (
          <TouchableOpacity
            key={opt}
            style={[styles.optionButton, current === opt && styles.optionButtonActive]}
            onPress={() => setter(opt)}
          >
            <Text style={[styles.optionText, current === opt && styles.optionTextActive]}>{opt}</Text>
          </TouchableOpacity>
        ))}
      </View>
    </View>
  );

  return (
    <ScrollView style={styles.container}>
      <View style={styles.header}>
        <Text style={styles.title}>Fiche Terrain - {type.toUpperCase()}</Text>
        <Text style={styles.subtitle}>ID : {object.code || object.fid || object.id}</Text>
      </View>

      {/* Fiche Technique */}
      <View style={styles.card}>
        <Text style={styles.cardTitle}>Fiche Technique</Text>
        {object.diametre && <Text style={styles.infoText}>• Diamètre : {object.diametre} mm</Text>}
        {object.longueur && <Text style={styles.infoText}>• Longueur : {object.longueur} m</Text>}
        {object.materiau && <Text style={styles.infoText}>• Matériau : {object.materiau}</Text>}
        {object.profondeur && <Text style={styles.infoText}>• Profondeur : {object.profondeur} m</Text>}
        {object.commune && <Text style={styles.infoText}>• Commune : {object.commune}</Text>}
        {object.nom_voie && <Text style={styles.infoText}>• Voie : {object.nom_voie}</Text>}
      </View>

      {/* Formulaire d'inspection */}
      <View style={styles.card}>
        <Text style={styles.cardTitle}>Observations terrain</Text>
        
        {renderDropdown('État général', etat, ['Bon', 'Moyen', 'Mauvais'], setEtat)}
        {renderDropdown("Niveau d'eau", niveauEau, ['Vide', 'Normal', 'Sous charge', 'Débordement'], setNiveauEau)}
        {renderDropdown('Obstruction', obstruction, ['Aucune', 'Partielle', 'Totale'], setObstruction)}
        {renderDropdown('Odeur', odeur, ['Aucune', 'Légère', 'Forte'], setOdeur)}

        <View style={styles.formRow}>
          <Text style={styles.label}>Débordement actif :</Text>
          <Switch value={debordement} onValueChange={setDebordement} trackColor={{ true: '#e74c3c' }} />
        </View>

        {renderDropdown('Statut de validation', statut, ['À vérifier', 'Corrigé terrain', 'Validé bureau'], setStatut)}

        <Text style={styles.label}>Commentaires / Diagnostic :</Text>
        <TextInput
          style={styles.textInput}
          multiline
          numberOfLines={3}
          value={commentaires}
          onChangeText={setCommentaires}
          placeholder="Renseigner les constatations terrain..."
        />

        {/* Photos d'inspection */}
        <Text style={styles.label}>Photos d'inspection ({photos.length})</Text>
        <View style={styles.photoGrid}>
          {photos.map((uri, idx) => (
            <Image key={idx} source={{ uri }} style={styles.thumbnail} />
          ))}
          <TouchableOpacity style={styles.addPhotoBtn} onPress={handleAddPhoto}>
            <Text style={styles.addPhotoText}>📸 + Photo</Text>
          </TouchableOpacity>
        </View>
      </View>

      <View style={styles.buttons}>
        <TouchableOpacity style={[styles.btn, styles.btnCancel]} onPress={onClose}>
          <Text style={styles.btnTextCancel}>Annuler</Text>
        </TouchableOpacity>
        <TouchableOpacity style={[styles.btn, styles.btnSave]} onPress={handleSave}>
          <Text style={styles.btnTextSave}>Enregistrer</Text>
        </TouchableOpacity>
      </View>
    </ScrollView>
  );
}

const styles = StyleSheet.create({
  container: { flex: 1, backgroundColor: '#f9f9f9', padding: 12 },
  header: { backgroundColor: '#16213e', padding: 16, borderRadius: 8, marginBottom: 12 },
  title: { color: '#fff', fontSize: 18, fontWeight: 'bold' },
  subtitle: { color: '#ddd', fontSize: 14, marginTop: 4 },
  card: { backgroundColor: '#fff', padding: 16, borderRadius: 8, marginBottom: 12, elevation: 1 },
  cardTitle: { fontSize: 15, fontWeight: 'bold', color: '#16213e', marginBottom: 10, borderBottomWidth: 1, borderBottomColor: '#eee', paddingBottom: 6 },
  infoText: { fontSize: 13, color: '#555', marginBottom: 4 },
  formGroup: { marginBottom: 12 },
  label: { fontSize: 13, fontWeight: 'bold', color: '#333', marginBottom: 6 },
  optionContainer: { flexDirection: 'row', flexWrap: 'wrap' },
  optionButton: { backgroundColor: '#f0f0f0', paddingHorizontal: 10, paddingVertical: 6, borderRadius: 4, marginRight: 6, marginBottom: 6 },
  optionButtonActive: { backgroundColor: '#16213e' },
  optionText: { fontSize: 12, color: '#555' },
  optionTextActive: { color: '#fff', fontWeight: 'bold' },
  formRow: { flexDirection: 'row', justifyContent: 'space-between', alignItems: 'center', marginVertical: 8 },
  textInput: { borderWidth: 1, borderColor: '#ddd', borderRadius: 4, padding: 8, fontSize: 13, backgroundColor: '#fafafa', textAlignVertical: 'top', minHeight: 60, marginBottom: 12 },
  photoGrid: { flexDirection: 'row', flexWrap: 'wrap', alignItems: 'center', marginTop: 6 },
  thumbnail: { width: 50, height: 50, borderRadius: 4, marginRight: 8, marginBottom: 8 },
  addPhotoBtn: { width: 50, height: 50, borderRadius: 4, borderWidth: 1, borderStyle: 'dashed', borderColor: '#888', justifyContent: 'center', alignItems: 'center', marginBottom: 8 },
  addPhotoText: { fontSize: 9, fontWeight: 'bold', color: '#555', textAlign: 'center' },
  buttons: { flexDirection: 'row', justifyContent: 'space-between', marginTop: 12, marginBottom: 40 },
  btn: { flex: 1, padding: 12, borderRadius: 6, alignItems: 'center', marginHorizontal: 4 },
  btnCancel: { backgroundColor: '#fff', borderWidth: 1, borderColor: '#888' },
  btnSave: { backgroundColor: '#27ae60' },
  btnTextCancel: { color: '#555', fontWeight: 'bold' },
  btnTextSave: { color: '#fff', fontWeight: 'bold' }
});
